keps/sig-node/3008-cri-class-based-resources/README.md

```diff
@@ -0,0 +1,1257 @@
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+This is only the first step in getting class-based resources supported in
+Kubernetes. Important pieces like resource assignment via pod spec, resource
+status, resource disovery and permission control are [non-goals](#non-goals)
+not solved here.
+These aspects
+are briefly discussed in [future work](#future-work). The risk in this sort of
+piecemeal approach is finding devil in the details, resulting in inconsistent
+and/or crippled and/or cumbersome end result. However, there is a lot of
+experience in extending the API and understanding which sort of solutions are
+functional and practical.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+- User assigning container to “unauthorized” class, causing interference and
+  access to unwanted set/amount of resources. This will be addressed in future
+  KEP introducing permission controls.
+- Confusion: user tries to assign container to RDT class but RDT has not been
+  enabled on system(s). This will be addressed by future KEP(s) introducing
+  resource discovery and status.
+- Keeping client (kubelet) and runtime in sync wrt to available classes. Will
+  be addressed in future KEP about resource discovery.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+Configuration and management of the class resources is fully handled by the
+underlying container runtime and is invisible to kubelet. An error to the CRI
+client is returned if the specified class is not available.
+
+### CRI protocol
+
+The following additions to the CRI protocol are suggested.
+
+The `ContainerConfig` message will be supplemented with new `class_resources`
+field, providing per-container setting for class resources.
+
+
+```diff
+ message ContainerConfig {
+
+ ...
+     // Configuration specific to Linux containers.
+     LinuxContainerConfig linux = 15;
+     // Configuration specific to Windows containers.
+     WindowsContainerConfig windows = 16;
++
++    // Configuration of class resources.
++    ContainerClassResources class_resources = 17;
+ }
+
++// ContainerClassResources specifies the configuration of class based
++// resources of a container.
++message ContainerClassResources {
++    // Resource classes of the container will be assigned to
++    map<string, string> classes = 1;
```

Comment:
in the proposal for pod changes resources are forming list. Here are resources are in the map. What is your thinking on which layer will in future validate the names uniquiness?

Reply:
Hmm, what list are you referring to? 🧐 The assignment should be map everywhere. In resource discovery/status there are lists. I think we could use only maps everywhere or only lists everywhere. The reason for me using map in the assignment part was to follow existing "style" with other resources. In resource discovery/status part I used lists to follow the api convention to prefer lists (of named subobjects) over maps.

Regarding validating duplicates, maps of course have that property by nature. If we would use lists in the Pod spec I think it would be the "standard" validation package. Regarding CRI API, I haven't really thought about that other that logging a warning if duplicates are detected (and using the last one).

Reply:
@SergeyKanzhelev the API(s) in the proposal have changed from using maps to lists. In case of Kubernetes api the apiserver (validation) will be responsible for that.